contracts/token/ERC721_Enumerable_base.cairo

```diff
@@ -0,0 +1,245 @@
+        pedersen_ptr: HashBuiltin*, 
+        syscall_ptr: felt*, 
+        range_check_ptr
+    }(to: felt, token_id: Uint256):
+    _add_token_to_all_tokens_enumeration(token_id)
+    _add_token_to_owner_enumeration(to, token_id)
+    ERC721_mint(to, token_id)
+    return ()
+end
+
+func ERC721_Enumerable_burn{
+        pedersen_ptr: HashBuiltin*, 
+        syscall_ptr: felt*, 
+        range_check_ptr
+    }(token_id: Uint256):
+    let (_from) = ERC721_ownerOf(token_id)
+    _remove_token_from_owner_enumeration(_from, token_id)
+    _remove_token_from_all_tokens_enumeration(token_id)
+    ERC721_burn(token_id)
+    return ()
+end
+
+func ERC721_Enumerable_transferFrom{
+        pedersen_ptr: HashBuiltin*, 
+        syscall_ptr: felt*, 
+        range_check_ptr
+    }(_from: felt, to: felt, token_id: Uint256):
+    _remove_token_from_owner_enumeration(_from, token_id)
+    _add_token_to_owner_enumeration(to, token_id)
+    ERC721_transferFrom(_from, to, token_id)
+    return ()
+end
+
+func ERC721_Enumerable_safeTransferFrom{
+        pedersen_ptr: HashBuiltin*, 
+        syscall_ptr: felt*, 
+        range_check_ptr
+    }(
+        _from: felt, 
+        to: felt, 
+        token_id: Uint256, 
+        data_len: felt,
+        data: felt*
+    ):
+    _remove_token_from_owner_enumeration(_from, token_id)
+    _add_token_to_owner_enumeration(to, token_id)
+    ERC721_safeTransferFrom(_from, to, token_id, data_len, data)
+    return ()
+end
+
+#
+# Internals
+#
+
+func _add_token_to_all_tokens_enumeration{
+        pedersen_ptr: HashBuiltin*, 
+        syscall_ptr: felt*, 
+        range_check_ptr
+    }(token_id: Uint256):
+    alloc_locals
+    let (supply: Uint256) = ERC721_Enumerable_all_tokens_len.read()
+    ERC721_Enumerable_all_tokens.write(supply, token_id)
+    ERC721_Enumerable_all_tokens_index.write(token_id, supply)
+    
+    let (local new_supply: Uint256, _) = uint256_add(supply, Uint256(1, 0))
+    ERC721_Enumerable_all_tokens_len.write(new_supply)
+    return ()
+end
+
+
+func _remove_token_from_all_tokens_enumeration{
+        pedersen_ptr: HashBuiltin*, 
+        syscall_ptr: felt*, 
+        range_check_ptr
+    }(token_id: Uint256):
+    alloc_locals
+    let (local supply: Uint256) = ERC721_Enumerable_all_tokens_len.read()
+    let (local last_token_index: Uint256) = uint256_sub(supply, Uint256(1, 0))
+    let (local token_index: Uint256) = ERC721_Enumerable_all_tokens_index.read(token_id)
+
+    # When the token to delete is the last token, the swap operation is unnecessary. However,
+    # since this occurs so rarely (when the last minted token is burnt), we still do the swap
+    # here to avoid the gas cost of adding an 'if' statement (like in _remove_token_from_owner_enumeration)
+    let (local last_token_id: Uint256) = ERC721_Enumerable_all_tokens.read(last_token_index)
+
+    ERC721_Enumerable_all_tokens.write(last_token_index, Uint256(0, 0))
+    ERC721_Enumerable_all_tokens.write(token_index, last_token_id)
```

Comment:
Hm correct me if I'm wrong but I think when `last_token_id` and `token_index` are equal (this is either when they're `0` or the last element) the first of this two lines would correctly delete it to `0` but then the next lines overwrites it with the just-deleted `last_token_id`'s previous value. I'm not sure if this is very important yet (*) since the value won't be readable thanks to the index/supply check on the getters.


(*) maybe in the future it could be fixed if it saves gas

Reply:
I believe you are correct. I initially did not have the `last_token_index` set to zero because of index/supply check and for the situation you described, but I changed my mind after your initial review to play it safe. I agree with putting this in the 'nice-to-have' queue especially if it saves gas